src/System.Private.CoreLib/src/System/GC.cs

```diff
@@ -508,5 +509,118 @@ public static void EndNoGCRegion()
         {
             EndNoGCRegionWorker();
         }
+
+        readonly struct MemoryLoadChangeNotification
+        {
+            public float LowMemoryPercent { get; }
+            public float HighMemoryPercent { get; }
+            public Action Notification { get; }
+
+            public MemoryLoadChangeNotification(float lowMemoryPercent, float highMemoryPercent, Action notification)
+            {
+                LowMemoryPercent = lowMemoryPercent;
+                HighMemoryPercent = highMemoryPercent;
+                Notification = notification;
+            }
+        }
+
+        private static List<MemoryLoadChangeNotification> s_notifications = new List<MemoryLoadChangeNotification>();
+        private static Action<float> s_notificationsCallback = new Action<float>(InvokeMemoryLoadChangeNotifications);
+        private static IntPtr s_notificationsFtnPtr = Marshal.GetFunctionPointerForDelegate(s_notificationsCallback);
+
+        private static void InvokeMemoryLoadChangeNotifications(float memoryPercent)
+        {
+            lock (s_notifications)
+            {
+                int last = 0;
+                for (int i = 0; i < s_notifications.Count; ++i)
+                {
+                    if (s_notifications[i].LowMemoryPercent <= memoryPercent && memoryPercent <= s_notifications[i].HighMemoryPercent)
+                    {
+                        s_notifications[i].Notification();
```

Comment:
If you are concerned of deadlocks between `InvokeMemoryLoadChangeNotifications` and `RegisterMemoryLoadChangeNotification`/`UnregisterMemoryLoadChangeNotification`, it shouldn't be possible for it to happen since the latters should not be taking anyother lock (it's only callong `List.RemoteAt` or `List.Add` which is safe to assume will not be taking any lock).

Are you concerned of something else I am missing?

Reply:
The `s_notificationsFtnPtr` is invoked from the finalizer, right?  And it invokes arbitrary user-provided delegates?

The finalizer is a single resource, the equivalent of holding a lock that ensures mutual exclusion of all other finalizers.  So if the finalizer invokes a callback that blocks waiting on a lock (some arbitrary user lock) that's held by some other thread that's doing something that'll block waiting for the finalizer to make progress, it'll deadlock, right?  e.g. GC.WaitForPendingFinalizers, or blocking waiting for a finalizable object to set some event, or some such thing?

If it's not a real concern, ok, but I get nervous any time I see code invoking arbitrary delegates while holding a lock.

Reply:
You are perfectly right! The documentation comes to some help here, with an explicit `The notification will occur on a thread which should not be blocked. Complex processing in the notification should defer work to the threadpool.` at https://github.com/dotnet/coreclr/pull/24202/files#diff-8e52c75a09516f4f3a78d2d6973fa364R594. I don't know how much we can usually expect consumers of the API to respect the assumptions stated in the API.